src/compiler/rust.rs

```diff
@@ -1023,18 +1187,88 @@ impl Compilation for RustCompilation {
     }
 
     #[cfg(feature = "dist-client")]
-    fn into_dist_packagers(self: Box<Self>, path_transformer: &mut dist::PathTransformer) -> Result<(Box<pkg::InputsPackager>, Box<pkg::ToolchainPackager>)> {
+    fn into_dist_packagers(self: Box<Self>, path_transformer: dist::PathTransformer) -> Result<(Box<pkg::InputsPackager>, Box<pkg::ToolchainPackager>, Box<OutputsRewriter>)> {
 
-        let RustCompilation { inputs, crate_link_paths, sysroot, crate_types, .. } = *{self};
+        let RustCompilation { inputs, crate_link_paths, sysroot, crate_types, dep_info, rlib_dep_reader, env_vars, .. } = *{self};
         trace!("Dist inputs: inputs={:?} crate_link_paths={:?}", inputs, crate_link_paths);
 
+        let inputs_packager = Box::new(RustInputsPackager { env_vars, crate_link_paths, crate_types, inputs, path_transformer, rlib_dep_reader });
+        let toolchain_packager = Box::new(RustToolchainPackager { sysroot });
+        let outputs_rewriter = Box::new(RustOutputsRewriter { dep_info });
+
+        Ok((inputs_packager, toolchain_packager, outputs_rewriter))
+    }
+
+    fn outputs<'a>(&'a self) -> Box<Iterator<Item=(&'a str, &'a Path)> + 'a> {
+        Box::new(self.outputs.iter().map(|(k, v)| (k.as_str(), &**v)))
+    }
+}
+
+// TODO: we do end up with slashes facing the wrong way, but Windows is agnostic so it's
+// mostly ok. We currently don't get mappings for every single path because it means we need to
+// figure out all prefixes and send them over the wire.
+#[cfg(feature = "dist-client")]
+fn get_path_mappings(path_transformer: &dist::PathTransformer) -> impl Iterator<Item=(PathBuf, String)> {
+    path_transformer.disk_mappings()
+}
+
+#[cfg(feature = "dist-client")]
+struct RustInputsPackager {
+    env_vars: Vec<(OsString, OsString)>,
+    crate_link_paths: Vec<PathBuf>,
+    crate_types: CrateTypes,
+    inputs: Vec<PathBuf>,
+    path_transformer: dist::PathTransformer,
+    rlib_dep_reader: Option<Arc<RlibDepReader>>,
+}
+
+#[cfg(feature = "dist-client")]
+impl pkg::InputsPackager for RustInputsPackager {
+    fn write_inputs(self: Box<Self>, wtr: &mut io::Write) -> Result<dist::PathTransformer> {
+        use ar;
+        use tar;
+
+        debug!("Packaging compile inputs for compile");
+        let RustInputsPackager { crate_link_paths, crate_types, inputs, mut path_transformer, rlib_dep_reader, env_vars } = *{self};
+
+        // If this is a cargo build, we can assume all immediate `extern crate` dependencies
+        // have been passed on the command line, allowing us to scan them all and find the
+        // complete list of crates we might need.
+        // If it's not a cargo build, we can't to extract the `extern crate` statements and
+        // so have no way to build a list of necessary crates - send all rlibs.
```

Comment:
It's nice that you have fallback support here, but it would have been fine to just not support distributed compilation for the non-cargo case, honestly.